### PR TITLE
Bug fix 1683359 - fluentd metrics not found

### PR DESCRIPTION
**Pull request by josefkarasek**

### CODE REVIEW

File: install.sh [hunk 1]
```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/env bash
```
Comment (jcantrill) [inline install.sh]: I assume this is not required? since we have make targets that have a number of these things?

File: pkg/utils/types.go [hunk 2]
```diff
@@ -348,3 +349,16 @@ func NewClusterRoleBinding(bindingName, roleName string, subjects []rbac.Subject
 		Subjects: subjects,
 	}
 }
+
+func ServiceMonitor(serviceMonitorName, namespace string) *monitoringv1.ServiceMonitor {
+	return &monitoringv1.ServiceMonitor{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       monitoringv1.ServiceMonitorsKind,
+			APIVersion: monitoringv1.Group + "/" + monitoringv1.Version,
```
Comment (jcantrill) [inline pkg/utils/types.go]: Can this not be similar to `APIVersion: batch.SchemeGroupVersion.String()` for SM?

File: install.sh (hunk 1 above)
Comment (richm) [reply install.sh]: not sure what you are asking here - you need the `#!/usr/bin/env` (could also use `#!/bin/bash`) in order for a shell script to be executable with the specified interpreter.

File: install.sh [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+echo 'vm.max_map_count=262144' | sudo tee --append /etc/sysctl.d/99-elasticsearch.conf
```
Comment (richm) [inline install.sh]: This only works if you are doing a local install - see REMOTE_CLUSTER - also, you have to apply it to every elasticsearch node in the cluster

File: install.sh [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+oc create -n openshift-logging -f https://raw.githubusercontent.com/coreos/prometheus-operator/master/example/prometheus-operator-crd/servicemonitor.crd.yaml
+oc create -n openshift-logging -f https://raw.githubusercontent.com/openshift/cluster-logging-operator/01cbb3c3c4fb3c34fc7204497747038c4ff22779/deploy/cr.yaml
+
+oc create -f hack/cr.yaml
+
+REPO_PREFIX=openshift/ IMAGE_PREFIX=origin- OPERATOR_NAME=cluster-logging-operator WATCH_NAMESPACE=openshift-logging KUBERNETES_CONFIG=~/.kube/config ELASTICSEARCH_IMAGE=docker.io/openshift/origin-logging-elasticsearch5 ./main
+
+ALERTS_FILE_PATH=files/prometheus_alerts.yml RULES_FILE_PATH=files/prometheus_rules.yml PROXY_IMAGE=quay.io/openshift/origin-oauth-proxy:v4.0.0 REPO_PREFIX=openshift/ IMAGE_PREFIX=origin- OPERATOR_NAME=elasticsearch-operator WATCH_NAMESPACE=openshift-logging KUBERNETES_CONFIG=~/.openshift/aws/auth/kubeconfig LOG_LEVEL=debug go run cmd/elasticsearch-operator/main.go
+
+docker.io/openshift/oauth-proxy:v1.1.0
```
Comment (richm) [inline install.sh]: ?

Comment (richm) [reply install.sh]: was the rest of this file intended to go into a readme or comment?

Comment (josefkarasek) [reply install.sh]: I committed the file accidentally - I will remove it.

**Comment by richm** [comment]
ok - test is failing for real, not a flake - please investigate

**Comment by nhosoi** [comment]
Slack/email conversation with @josefkarasek:

Regarding this question:
> When you deploy fluentd, can you rsh inside and hit the metrics endpoint? My PR isnt merged because my fluentd container somehow doesn't expose metrics

you suggested me to run the following command in the fluentd pod:
> curl localhost:24224/metrics

It failed, but "curl --insecure https://localhost:24231/metrics" returns the expected fluentd metrics both in the fluentd pod.

Looking at the pkg/k8shandler/collection.go file in your pr/112, the port (== 24231) and the path (/metrics) match the default fluentd prometheus config params.  So, can we say we could "rsh inside and hit the metrics endpoint"?

The other thing to check could be the cert?
```
    prometheusCAFile  = "/etc/prometheus/configmaps/serving-certs-ca-bundle/service-ca.crt"
        TLSConfig: &monitoringv1.TLSConfig{
            CAFile:     prometheusCAFile,
            ServerName: fmt.Sprintf("%s.%s.svc", "fluentd", cluster.Namespace),
            // ServerName can be e.g. fluentd.openshift-logging.svc
        },
```

**Comment by nhosoi** [comment]
Hi @josefkarasek, a question, actually two. :)

This line [1] shows /etc/fluent/metrics is supposed be host mounted for the secret.  But it's not found in my fluentd pods - [2]...
[1] - https://github.com/josefkarasek/cluster-logging-operator/blob/bz1683359-fluentd-metrics/pkg/k8shandler/collection.go#L623

[2]
```
$ oc exec fluentd-8t5z2 -n openshift-logging -- ls /etc/fluent/metrics
ls: cannot access /etc/fluent/metrics: No such file or directory
command terminated with exit code 2
$ oc exec fluentd-8t5z2 -n openshift-logging -- ls /etc/fluent
configs.d
fluent.conf
keys
plugin
```
And it seems we are creating a service for metrics (at least elasticsearch has one elasticsearch-metrics), but there's no one for fluentd.  Is it ok?
```
$ oc get service -n openshift-logging
NAME                    TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)     AGE
elasticsearch           ClusterIP   172.30.206.250   <none>        9200/TCP    174m
elasticsearch-cluster   ClusterIP   172.30.130.45    <none>        9300/TCP    174m
elasticsearch-metrics   ClusterIP   172.30.203.66    <none>        60000/TCP   174m
kibana                  ClusterIP   172.30.162.155   <none>        443/TCP     174m
```

**Comment by nhosoi** [comment]
I have a new files/fluentd_prometheus_alerts.yaml to be replaced with the one in this pr.
https://github.com/nhosoi/cluster-logging-operator/blob/pr_112/files/fluentd_prometheus_alerts.yaml
Since this pr had a hard time to pass all the tests, I con't mind submitting a new pr just for the file.  I'll leave the decision to you, @josefkarasek.

File: pkg/k8shandler/monitoring.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+		return nil, err
+	}
+
+	rule := utils.PrometheusRule(ruleName, namespace)
+	rule.Spec = *alertsRuleSpec
+
+	return rule, nil
+}
+
+func ruleSpec(filePath string) (*monitoringv1.PrometheusRuleSpec, error) {
```
Comment (jcantrill) [inline pkg/k8shandler/monitoring.go]: Consider using the name `newPrometheusRuleSpecFrom(filepath)`.

File: pkg/k8shandler/monitoring.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+		return nil, fmt.Errorf("'%s' not readable", filePath)
+	}
+	ruleSpec := monitoringv1.PrometheusRuleSpec{}
+	if err := k8sYAML.NewYAMLOrJSONDecoder(bytes.NewBufferString(string(fileContent)), 1000).Decode(&ruleSpec); err != nil {
+		return nil, err
+	}
+	return &ruleSpec, nil
+}
+
+func checkFile(filePath string) error {
```
Comment (jcantrill) [inline pkg/k8shandler/monitoring.go]: `checkFileExists` or `verifyFileExists`

File: pkg/utils/factory.go [hunk 7]
```diff
@@ -363,3 +364,29 @@ func NewClusterRoleBinding(bindingName, roleName string, subjects []rbac.Subject
 		Subjects: subjects,
 	}
 }
+
+func ServiceMonitor(serviceMonitorName, namespace string) *monitoringv1.ServiceMonitor {
```
Comment (jcantrill) [inline pkg/utils/factory.go]: `NewServiceMonitor`

File: pkg/utils/factory.go [hunk 8]
```diff
@@ -363,3 +364,29 @@ func NewClusterRoleBinding(bindingName, roleName string, subjects []rbac.Subject
...
+			APIVersion: monitoringv1.Group + "/" + monitoringv1.Version,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      serviceMonitorName,
+			Namespace: namespace,
+		},
+	}
+}
+
+func PrometheusRule(ruleName, namespace string) *monitoringv1.PrometheusRule {
```
Comment (jcantrill) [inline pkg/utils/factory.go]: `NewPrometheusRule`

File: pkg/utils/factory.go [hunk 9]
```diff
@@ -363,3 +364,29 @@ func NewClusterRoleBinding(bindingName, roleName string, subjects []rbac.Subject
 		Subjects: subjects,
 	}
 }
+
+func ServiceMonitor(serviceMonitorName, namespace string) *monitoringv1.ServiceMonitor {
+	return &monitoringv1.ServiceMonitor{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       monitoringv1.ServiceMonitorsKind,
+			APIVersion: monitoringv1.Group + "/" + monitoringv1.Version,
```
Comment (jcantrill) [inline pkg/utils/factory.go]: Can this be replaced with `monitoringv1.SchemeGroupVersion.String()` Ref: https://github.com/openshift/cluster-logging-operator/blob/master/pkg/utils/factory.go#L21

File: manifests/03-role.yaml [hunk 10]
```diff
@@ -78,6 +78,13 @@ rules:
   - "privileged"
   verbs:
   - use
+- apiGroups:
```
Comment (richm) [inline manifests/03-role.yaml]: @jcantrill does this need to go into the CSV?

File: hack/cr.yaml [hunk 11]
```diff
@@ -10,6 +10,7 @@ spec:
       nodeCount: 1
       storage: {}
       redundancyPolicy: "ZeroRedundancy"
+      resources: {}
```
Comment (richm) [inline hack/cr.yaml]: ?

Comment (jcantrill) [reply hack/cr.yaml]: Doesn't hurt but doesn't add anything

File: manifests/03-role.yaml (hunk 10 above)
Comment (jcantrill) [reply manifests/03-role.yaml]: Correct.  This change does not affect the actual deployment and needs to go here for certain: https://github.com/openshift/cluster-logging-operator/blob/master/controller-manifests/cluster-logging.v4.1.0.clusterserviceversion.yaml and possibly here temporarily https://github.com/operator-framework/community-operators/blob/master/community-operators/cluster-logging/cluster-logging.v0.0.1.clusterserviceversion.yaml

**Comment by nhosoi** [comment]
I observed the elasticsearch-operator crash:
`elasticsearch-operator-54f5cc446f-dnlhg              0/1     Error     14         3h5m`

This might be a transient phenomenon since I'm still using the source code based on this PR, which would be a bit behind the latest master.  

Plus, it automatically restarted as follows.  But as you see, it was "restarted" 15 times by now...
`elasticsearch-operator-54f5cc446f-dnlhg              1/1     Running   15         3h24m`

This is the test command line that caused the crash.
while true; do oc delete pod $( oc get pod -l component=elasticsearch | egrep elasticsearch | awk '{print $1}' ); done

elasticsearch-operator log when the elasticsearch-operator is in the error state:
```
$ oc logs elasticsearch-operator-54f5cc446f-dnlhg
time="2019-03-26T03:39:13Z" level=info msg="Go Version: go1.10.3"
time="2019-03-26T03:39:13Z" level=info msg="Go OS/Arch: linux/amd64"
time="2019-03-26T03:39:13Z" level=info msg="operator-sdk Version: 0.0.7"
time="2019-03-26T03:39:13Z" level=info msg="Watching logging.openshift.io/v1alpha1, Elasticsearch, , 5000000000"
time="2019-03-26T03:39:21Z" level=info msg="Waiting for cluster to be fully recovered before restarting elasticsearch-clientdatamaster-0-1:  / green"
time="2019-03-26T03:39:39Z" level=info msg="Waiting for cluster to be fully recovered before restarting elasticsearch-clientdatamaster-0-1:  / green"
E0326 03:39:48.493584       1 runtime.go:66] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go/src/github.com/openshift/elasticsearch-operator/_output/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go:72
/go/src/github.com/openshift/elasticsearch-operator/_output/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/go/src/github.com/openshift/elasticsearch-operator/_output/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:573
/usr/local/go/src/runtime/panic.go:502
/usr/local/go/src/runtime/panic.go:63
/usr/local/go/src/runtime/signal_unix.go:388
/go/src/github.com/openshift/elasticsearch-operator/_output/src/github.com/openshift/elasticsearch-operator/pkg/k8shandler/status.go:211
/go/src/github.com/openshift/elasticsearch-operator/_output/src/github.com/openshift/elasticsearch-operator/pkg/k8shandler/status.go:48
/go/src/github.com/openshift/elasticsearch-operator/_output/src/github.com/openshift/elasticsearch-operator/pkg/k8shandler/cluster.go:171
/go/src/github.com/openshift/elasticsearch-operator/_output/src/github.com/openshift/elasticsearch-operator/pkg/stub/handler.go:67
/go/src/github.com/openshift/elasticsearch-operator/_output/src/github.com/openshift/elasticsearch-operator/pkg/stub/handler.go:29
/go/src/github.com/openshift/elasticsearch-operator/_output/src/github.com/operator-framework/operator-sdk/pkg/sdk/informer-sync.go:88
/go/src/github.com/openshift/elasticsearch-operator/_output/src/github.com/operator-framework/operator-sdk/pkg/sdk/informer-sync.go:52
/go/src/github.com/openshift/elasticsearch-operator/_output/src/github.com/operator-framework/operator-sdk/pkg/sdk/informer-sync.go:36
/go/src/github.com/openshift/elasticsearch-operator/_output/src/github.com/operator-framework/operator-sdk/pkg/sdk/informer.go:98
/go/src/github.com/openshift/elasticsearch-operator/_output/src/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/go/src/github.com/openshift/elasticsearch-operator/_output/src/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/go/src/github.com/openshift/elasticsearch-operator/_output/src/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/usr/local/go/src/runtime/asm_amd64.s:2361
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xf8bee8]

goroutine 155 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/github.com/openshift/elasticsearch-operator/_output/src/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x107
panic(0x10a5220, 0x1984bd0)
	/usr/local/go/src/runtime/panic.go:502 +0x229
github.com/openshift/elasticsearch-operator/pkg/k8shandler.updateNodeConditions(0xc42096d3f0, 0xd, 0xc42080faa0, 0x11, 0xc420848420)
	/go/src/github.com/openshift/elasticsearch-operator/_output/src/github.com/openshift/elasticsearch-operator/pkg/k8shandler/status.go:211 +0x5b8
github.com/openshift/elasticsearch-operator/pkg/k8shandler.UpdateClusterStatus(0xc4200e9500, 0xc4208c2480, 0x1)
	/go/src/github.com/openshift/elasticsearch-operator/_output/src/github.com/openshift/elasticsearch-operator/pkg/k8shandler/status.go:48 +0x1e8
github.com/openshift/elasticsearch-operator/pkg/k8shandler.CreateOrUpdateElasticsearchCluster(0xc4200e9500, 0x0, 0x0)
	/go/src/github.com/openshift/elasticsearch-operator/_output/src/github.com/openshift/elasticsearch-operator/pkg/k8shandler/cluster.go:171 +0x666
github.com/openshift/elasticsearch-operator/pkg/stub.Reconcile(0xc4200e9500, 0xc4208a7c00, 0x12f1240)
	/go/src/github.com/openshift/elasticsearch-operator/_output/src/github.com/openshift/elasticsearch-operator/pkg/stub/handler.go:67 +0x2cf
github.com/openshift/elasticsearch-operator/pkg/stub.(*Handler).Handle(0x19b7ac8, 0x13057e0, 0xc4200b6018, 0x12f1240, 0xc4200e9500, 0xc4201b7d00, 0x0, 0x0)
	/go/src/github.com/openshift/elasticsearch-operator/_output/src/github.com/openshift/elasticsearch-operator/pkg/stub/handler.go:29 +0x76
github.com/operator-framework/operator-sdk/pkg/sdk.(*informer).sync(0xc420389ea0, 0xc4201e2760, 0x1f, 0x104a800, 0xc42042c250)
	/go/src/github.com/openshift/elasticsearch-operator/_output/src/github.com/operator-framework/operator-sdk/pkg/sdk/informer-sync.go:88 +0x12d
github.com/operator-framework/operator-sdk/pkg/sdk.(*informer).processNextItem(0xc420389ea0, 0xc4202f0200)
	/go/src/github.com/openshift/elasticsearch-operator/_output/src/github.com/operator-framework/operator-sdk/pkg/sdk/informer-sync.go:52 +0xd2
github.com/operator-framework/operator-sdk/pkg/sdk.(*informer).runWorker(0xc420389ea0)
	/go/src/github.com/openshift/elasticsearch-operator/_output/src/github.com/operator-framework/operator-sdk/pkg/sdk/informer-sync.go:36 +0x2b
github.com/operator-framework/operator-sdk/pkg/sdk.(*informer).(github.com/operator-framework/operator-sdk/pkg/sdk.runWorker)-fm()
	/go/src/github.com/openshift/elasticsearch-operator/_output/src/github.com/operator-framework/operator-sdk/pkg/sdk/informer.go:98 +0x2a
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc42042c2f0)
	/go/src/github.com/openshift/elasticsearch-operator/_output/src/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc42042c2f0, 0x3b9aca00, 0x0, 0x1, 0x0)
	/go/src/github.com/openshift/elasticsearch-operator/_output/src/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbd
k8s.io/apimachinery/pkg/util/wait.Until(0xc42042c2f0, 0x3b9aca00, 0x0)
	/go/src/github.com/openshift/elasticsearch-operator/_output/src/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by github.com/operator-framework/operator-sdk/pkg/sdk.(*informer).Run
	/go/src/github.com/openshift/elasticsearch-operator/_output/src/github.com/operator-framework/operator-sdk/pkg/sdk/informer.go:98 +0x209
```
@josefkarasek, could you please rebase this PR?  I'd expect it includes the elasticsearch rebase, too.  I'd like to test with the latest code...
Thanks!

**Comment by ewolinetz** [comment]
@nhosoi that shouldn't have anything to do with rebasing code here. that is a miss from that code... 
These are referencing the wrong spec field https://github.com/openshift/elasticsearch-operator/blob/master/pkg/k8shandler/status.go#L211-L212

I'll open up a follow up

File: pkg/utils/types.go (hunk 2 above)
Comment (nhosoi) [reply pkg/utils/types.go]: I could not find SchemeGroupVersion for "github.com/coreos/prometheus-operator/pkg/client/monitoring/v1".   It is available for "github.com/coreos/prometheus-operator/pkg/apis/monitoring/v1", though.  But I could not use it in monitoring.go...

**Comment by nhosoi** [comment]
Closing this pr in favor of https://github.com/openshift/cluster-logging-operator/pull/137

**Pull request closed without merging**